File: cs440/eindex.cpp
#include <string>
#include <iostream>
#include <vector>
#include <fstream>
#include <sstream>
#include <iterator>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

using namespace std;

class Employee
{
public:
    string name;
    string bio;
    string manager_id;
    string id;

    Employee() {}

    Employee(string id, string name, string bio, string manager_id)
    {
        this->id = id;
        this->name = name;
        this->bio = bio;
        this->manager_id = manager_id;
    }
};

size_t calc_hash(string thing)
{
    hash<string> hasher;
    return hasher(thing);
}

size_t calc_hash_emp(Employee emp)
{
    return calc_hash(emp.id);
}

class IndexEntry
{
    Employee emp;
    unsigned int bucket;
    bool valid;

public:
    IndexEntry(Employee emp, unsigned int bucket)
    {
        this->emp = emp;
        this->bucket = bucket;
        this->valid = true;
    }

    IndexEntry()
    {
        this->valid = false;
    }

    unsigned int get_bucket()
    {
        return this->bucket;
    }

    Employee get_employee()
    {
        return this->emp;
    }

    bool is_valid()
    {
        return this->valid;
    }
};

class EmployeeIndex
{
    vector<vector<Employee>> map;
    int buckets;
    bool valid = false;

public:
    EmployeeIndex()
    {
        this->buckets = 1;
        this->map = vector<vector<Employee>>();
        this->map.resize(1);
        this->map[0].resize(4);
    }

    EmployeeIndex(int buckets)
    {
        this->buckets = buckets;
        this->map = vector<vector<Employee>>();
        this->map.resize(buckets);
        for (int i = 0; i < this->buckets; i++)
        {
            map[i].resize(4);
        }
    }

    EmployeeIndex(vector<IndexEntry> entries)
    {
        unsigned int max_bucket = 1;
        for (int i = 0; i < entries.size(); i++)
        {
            if (entries[i].get_bucket() > max_bucket)
            {
                max_bucket = entries[i].get_bucket();
            }
        }

        vector<vector<Employee>> new_map = vector<vector<Employee>>();
        new_map.resize(max_bucket);
        for (int i = 0; i < entries.size(); i++)
        {
            new_map[entries[i].get_bucket()].push_back(entries[i].get_employee());
        }

        this->map = new_map;
        this->buckets = max_bucket;
        this->valid = true;
    }

    void insert(Employee emp)
    {
        // cout << "Inserting: " << emp.id << " " << emp.name << " " << emp.bio << " " << emp.manager_id << endl;
        this->valid = true;
        size_t hash = calc_hash_emp(emp);
        unsigned int bucket_idx = hash % this->buckets;
        if (this->map[bucket_idx].size() < 4)
        {
            this->map[bucket_idx].push_back(emp);
        }
        else
        {
            grow_and_insert(emp);
        }
    }

    void grow_and_insert(Employee emp)
    {
        // cout << "Growing and Inserting: " << emp.id << " " << emp.name << " " << emp.bio << " " << emp.manager_id << endl;
        vector<Employee> all_employees = vector<Employee>();
        for (int i = 0; i < this->map.size(); i++)
        {
            for (int j = 0; j < this->map[i].size(); j++)
            {
                // cout << "Extracting '" << this->map[i][j].id << "'" << endl;
                all_employees.push_back(this->map[i][j]);
            }
        }

        // cout << "Extracted all employees from map." << endl;

        vector<vector<Employee>> new_map = vector<vector<Employee>>();
        for (int i = 0; i < all_employees.size(); i++)
        {
            Employee emp = all_employees[i];
            size_t hash = calc_hash_emp(emp);
            unsigned int bucket_idx = hash % this->buckets;
            if (this->map[bucket_idx].size() < 4)
            {
                this->map[bucket_idx].push_back(emp);
            }
        }

        this->valid = true;
        this->buckets++;
    }

    void set_map(EmployeeIndex emp)
    {
        this->valid = emp.is_valid();
        this->map = emp.get_map();
    }

    vector<vector<Employee>> get_map()
    {
        return this->map;
    }

    vector<IndexEntry> into_tuple_list()
    {
        vector<IndexEntry> ent = vector<IndexEntry>();
        for (int i = 0; i < this->map.size(); i++)
        {
            for (int j = 0; j < this->map[i].size(); j++)
            {
                IndexEntry entry = IndexEntry(this->map[i][j], i);
                ent.push_back(entry);
            }
        }

        return ent;
    }

    IndexEntry lookup(string lookupid)
    {
        size_t strhash = calc_hash(lookupid);

        unsigned int bucket_idx = strhash % this->buckets;
        for (int i = 0; i < this->map[bucket_idx].size(); i++)
        {
            if (this->map[bucket_idx][i].id == lookupid)
            {
                return IndexEntry(this->map[bucket_idx][i], i);
            };
        }

        return IndexEntry(); // Invalid version
    }

    bool is_valid()
    {
        return this->valid;
    }
};


void write_index(EmployeeIndex index)
{
    vector<IndexEntry> tuples = index.into_tuple_list();
    ofstream file("./EmployeeIndex");

    for (int i = 0; i < tuples.size(); i++)
    {
        Employee emp = tuples[i].get_employee();
        file << to_string(tuples[i].get_bucket()) << "|" << emp.id << "," << emp.name << "," << emp.bio << "," << emp.manager_id << endl;
    }

    file.close();
}

EmployeeIndex read_index(string path)
{
    ifstream file(path);
    string line;

    vector<IndexEntry> entries;

    while (getline(file, line))
    {
        stringstream ss(line);
        string bucket_str;
        getline(ss, bucket_str, '|');
        unsigned int bucket_id = stoi(bucket_str);

        string id;
        getline(ss, id, ',');

        string name;
        getline(ss, name, ',');

        string bio;
        getline(ss, bio, ',');

        string manager_id;
        getline(ss, manager_id);

        entries.push_back(IndexEntry(Employee(id, name, bio, manager_id), bucket_id));
    }

    EmployeeIndex index = EmployeeIndex(entries);
}

EmployeeIndex read_csv(string path)
{
    ifstream file(path);
    if (file.fail()) {
        return EmployeeIndex();
    }
    cout << "Loaded file from path: " << path << endl;

    vector<Employee> entries;
    string line;

    while (getline(file, line))
    {
        string copied_line;

        cout << "Got line: " << line << endl;
        if (line.length() == 0) { break; }
        stringstream ss(line);

        string id;
        getline(ss, id, ',');

        string name;
        getline(ss, name, ',');

        string bio;
        getline(ss, bio, ',');

        string manager_id;
        getline(ss, manager_id);


        entries.push_back(Employee(id, name, bio, manager_id));
    }

    for (int i = 0; i < entries.size(); i++)
    {
        cout << "ID: '" << entries[i].id << "'" <<  endl;
    }

    cout << "All lines processed." << endl;

    EmployeeIndex index = EmployeeIndex();
    for (int i = 0; i < entries.size(); i++)
    {
        index.insert(entries[i]);
    }

    for (int i = 0; i < index.get_map().size() ; i++) {
        for (int j = 0; i < index.get_map()[i].size() ; i++) {
            cout << "Index in Map: " << index.get_map()[i][j].id << " " << index.get_map()[i][j].name << " " << index.get_map()[i][j].bio << " " << index.get_map()[i][j].manager_id << endl;
        }
    }

    return index;
}


int main(int argc, char *argv[])
{
    //cout << argc << endl;
    //for (int i = 0; i < argc; i++) {
    //	cout << i << ":\t'" << argv[i] << "'" << endl;
    //}

    if (argc == 2 && strcmp(argv[1], "C") == 0)
    {
        cout << "Loading from CSV and creating the index." << endl;
        EmployeeIndex indx = read_csv("./Employee.csv");
        cout << "Please check for EmployeeIndex file in the current directory." << endl;
        write_index(indx);
        return 0;
    }

    if (argc == 3 && strcmp(argv[1], "L") == 0) {

        EmployeeIndex indx = read_index("./EmployeeIndex");
        if (!indx.is_valid()) {
            indx = read_csv("./Employee.csv");
            write_index(indx);
        }

        if (!indx.is_valid()) {
            cout << "Had an error reading the index, please try re-creating or checking Employee.csv and EmployeeIndex" << endl;
            return 2;
        }

        IndexEntry lookup = indx.lookup(argv[2]);
        if (!lookup.is_valid()) {
            cout << "Couldn't find entry with hash " << argv[2] << "." << endl;
            return 3;
        }

        Employee emp = lookup.get_employee();
        cout << "Entry:" << endl << "\tid: " << emp.id << endl << "\tname: " << emp.name << endl << "\tbio: " << emp.bio << endl << "\tmanager_id" << emp.manager_id << endl;
        return 0;
    }

    cout << "Usage:" << endl << "\teindex C" << endl << "\teindex L [hash_lookup]" << endl;

    return 4;
}

File: cs440/eindex.rs
#![allow(unused)]
use std::collections::hash_map::DefaultHasher;
use std::ffi::{CString, CStr};
use std::fs::File;
use std::hash::{Hash, Hasher};
use std::io::{BufRead, BufReader, Write};
use std::env::args;

#[repr(C)]
#[derive(Debug, Clone)]
pub struct Employee {
    id: CString,         // exactly 8 bytes
    name: CString,       // at most 200 bytes
    bio: CString,        // at most 500 bytes
    manager_id: CString, // exactly 8 bytes
}

impl Employee {
    fn from_id(id: &str) -> Self {
        Self {
            id: CString::new(id.as_bytes()).expect("An invalid id was passed to Employee"),
            name: CString::default(),
            bio: CString::default(),
            manager_id: CString::default(),
        }
    }
}

fn main() -> std::result::Result<(), Box<dyn std::error::Error>> {
    let mut args = args().skip(1);
    let control_arg = args.next();
    match control_arg.as_ref().map(|s| s.as_str()) {
        Some("C")  => {
            println!("Loading from CSV and creating the index.\nPlease check for EmployeeIndex file in the current directory.");

            let employees = read_csv("Employee.csv")?;

            let mut map = EmployeeMap::new();
            for emp in employees {
                map.insert(emp);
            }

            write_index("./EmployeeIndex", map)?;
        },
        Some("L") => {
            let employee_map = read_index("./EmployeeIndex");
            if employee_map.is_err() {
                let employees = read_csv("./Employee.csv")?;

                let mut map = EmployeeMap::new();
                for emp in employees {
                    map.insert(emp);
                }

                write_index("./EmployeeIndex", map)?;
            }

            if let Ok(employee_map) = read_index("./EmployeeIndex") {
                match args.next() {
                    Some(id) => {
                        if let Some(emp) = employee_map.lookup(&id) {
                            println!("Found Employee:\n{:#?}", emp);
                        } else {
                            eprintln!("Couldn't find employee with that ID.");
                        }
                    },
                    None => {
                        eprintln!("You must include an item ID to lookup. Usage: eindex L [id]");
                    }
                }
            } else {
                println!("Error reading from index, was it corrupted?");
            }
        },
        _ => {
            println!("Usage:");
            println!("\teindex C \tRead the input CSV and create the index file.");
            println!("\teindex L [id] \tRead the index file and lookup a record. Will create the index file if it doesn't exist.");
        }
    }

    Ok(())
}

impl Hash for Employee {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.id.hash(state);
    }
}

fn calc_hash<T: Hash>(t: &T) -> u64 {
    let mut s = DefaultHasher::new();
    t.hash(&mut s);
    s.finish()
}

#[derive(Debug)]
struct EmployeeMap {
    map: Vec<Vec<Employee>>,
    buckets: usize,
}

impl EmployeeMap {
    fn new() -> Self {
        Self {
            map: vec![vec![]; 1],
            buckets: 1,
        }
    }

    fn insert(&mut self, item: Employee) {
        let hash = calc_hash(&item);
        let bucket_idx = hash as usize % self.buckets;
        if self.map[bucket_idx].len() < 4 {
            self.map[bucket_idx].push(item);
        } else {
            self.grow_and_insert(item);
        }
    }

    fn grow_and_insert(&mut self, item: Employee) {
        let mut employees: Vec<Employee> = Vec::new();
        for bucket in self.map.iter().cloned() {
            for emp in bucket {
                employees.push(emp);
            }
        }

        employees.push(item);

        self.buckets += 1;
        let mut new_map: Vec<Vec<Employee>> = vec![vec![]; self.buckets];

        for employee in employees {
            let hash = calc_hash(&employee);
            let bucket_idx = hash as usize % self.buckets;
            if new_map[bucket_idx].len() < 4 {
                new_map[bucket_idx].push(employee);
            }
        }

        self.map = new_map;
    }

    fn into_tuple_list(self) -> Vec<(usize, Employee)> {
        let mut list: Vec<(usize, Employee)> = vec![];
        for (idx, bucket) in self.map.into_iter().enumerate() {
            for emp in bucket {
                list.push((idx, emp));
            }
        }
        list
    }

    fn from_tuple_list(list: Vec<(usize, Employee)>) -> Self {
        let mut max_bucket_size = 0;
        for item in &list {
            if item.0 > max_bucket_size {
                max_bucket_size = item.0;
            }
        }

        // We record bucket indicies in the file so the size is 1 + the max index
        max_bucket_size += 1;

        let mut map: Vec<Vec<Employee>> = vec![vec![]; max_bucket_size];
        for item in list {
            map[item.0].push(item.1);
        }

        EmployeeMap {
            map,
            buckets: max_bucket_size,
        }
    }

    fn lookup(&self, id: &str) -> Option<Employee> {
        let hash = calc_hash(&Employee::from_id(id));
        let bucket_idx = hash as usize % self.buckets;
        for item in &self.map[bucket_idx] {
            if item.id.to_str().unwrap() == id {
                return Some(item.clone());
            }
        }

        None
    }
}

fn read_index(filepath: &str) -> Result<EmployeeMap, std::io::Error> {
    let tuples = BufReader::new(File::open(filepath)?)
        .lines()
        .filter_map(Result::ok)
        .map(|line| {
            let mut tuple_parts = line.split("|");
            let bucket_idx = tuple_parts
                .next()
                .expect("Unknown format found in index file, please recreate.");
            let bucket_idx = bucket_idx
                .parse::<usize>()
                .expect("Unknown format found in index file, please recreate.");
            let employee = tuple_parts
                .next()
                .expect("Unknown format found in index file, please recreate.");
            let employee = parse_employee(employee)
                .expect("Unknown format found in index file, please recreate.");
            (bucket_idx, employee)
        })
        .collect::<Vec<_>>();

    Ok(EmployeeMap::from_tuple_list(tuples))
}

fn write_index(filepath: &str, map: EmployeeMap) -> Result<(), std::io::Error> {
    let mut file = File::create(filepath)?;
    let tuples = map.into_tuple_list();
    for (bucket, employee) in tuples {
        file.write(format!("{}|", bucket).as_bytes())?;
        file.write(employee.id.as_bytes())?;
        file.write(b",");
        file.write(employee.name.as_bytes())?;
        file.write(b",");
        file.write(employee.bio.as_bytes())?;
        file.write(b",");
        file.write(employee.manager_id.as_bytes())?;
        file.write(b"\n");
    }

    Ok(())
}

fn parse_employee(line: &str) -> Option<Employee> {
    let mut parts = line
        .split(",")
        .filter_map(|part| CString::new(part).ok())
        .collect::<Vec<CString>>();

    if parts[0].as_bytes().len() != 8
        || parts[3].as_bytes().len() != 8
        || parts[1].as_bytes().len() > 200
        || parts[2].as_bytes().len() > 500
    {
        println!("Invalid lengths for employee {:?}", parts);
        return None;
    }

    Some(Employee {
        id: parts[0].clone(),
        name: parts[1].clone(),
        bio: parts[2].clone(),
        manager_id: parts[3].clone(),
    })
}

fn read_csv(filepath: &str) -> Result<Vec<Employee>, std::io::Error> {
    let employees = BufReader::new(File::open(filepath)?)
        .lines()
        .filter_map(Result::ok)
        .filter(|line| line.len() > 0)
        .filter_map(|line| parse_employee(&line))
        .collect::<Vec<_>>();

    Ok(employees)
}
